program/src/processor.rs

```diff
@@ -510,84 +507,179 @@ impl Processor {
+                }
+
+                (false, None)
+            }
+            StakeStateV2::RewardsPool => return Err(ProgramError::InvalidAccountData),
+        };
+
+        // special case: for a full split, we only care that the destination becomes a valid stake account
+        // this prevents state changes in exceptional cases where a once-valid source has become invalid
+        // relocate lamports, copy data, and close the original account
+        if split_lamports == source_lamport_balance {
+            let mut destination_stake_state = source_stake_state;
+            let delegation = match (&mut destination_stake_state, option_dest_meta) {
+                (StakeStateV2::Stake(meta, stake, _), Some(dest_meta)) => {
+                    *meta = dest_meta;
+
+                    if is_active_or_activating {
+                        stake.delegation.stake
                     } else {
-                        // Otherwise, the new split stake should reflect the entire split
-                        // requested, less any lamports needed to cover the
-                        // split_rent_exempt_reserve.
-                        if source_stake.delegation.stake.saturating_sub(split_lamports)
-                            < minimum_delegation
-                        {
-                            return Err(StakeError::InsufficientDelegation.into());
-                        }
-
-                        (
-                            split_lamports,
-                            split_lamports.saturating_sub(
-                                validated_split_info
-                                    .destination_rent_exempt_reserve
-                                    .saturating_sub(destination_lamport_balance),
-                            ),
-                        )
-                    };
-
-                if split_stake_amount < minimum_delegation {
+                        0
+                    }
+                }
+                (StakeStateV2::Initialized(meta), Some(dest_meta)) => {
+                    *meta = dest_meta;
+
+                    0
+                }
+                (StakeStateV2::Uninitialized, None) => 0,
+                _ => unreachable!(),
+            };
+
+            if destination_lamport_balance
+                .saturating_add(split_lamports)
+                .saturating_sub(delegation)
+                < destination_rent_exempt_reserve
+            {
+                return Err(ProgramError::InsufficientFunds);
+            }
+
+            if is_active_or_activating && delegation < minimum_delegation {
+                return Err(StakeError::InsufficientDelegation.into());
+            }
+
+            set_stake_state(destination_stake_account_info, &destination_stake_state)?;
+            if source_stake_account_info.key != destination_stake_account_info.key {
+                source_stake_account_info.resize(0)?;
+            }
```

Comment:
`source_stake_account_info.key == destination_stake_account_info.key` is a strange case. Should we just guard above to throw an err in that case? Can't think of a reason why someone would do that.

Reply:
i dunno why we support it either tbh, other than that we always have and that there are tests for it which indicate it was on purpose. it could be that the original authors just had a preference for permissiveness

Reply:
Should we just kill this? It feels more like a potential footgun hanging around.